torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,395 @@
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr || n->kind() == prim::GetAttr) {
+          // TODO: handle interface attributes. For now, Exit if Module uses
+          // inteface attributes
+          if (n->kind() == prim::GetAttr) {
+            TORCH_CHECK(
+                !n->output()->type()->cast<InterfaceType>(),
+                "attempted to freeze a module that uses interface attributes");
+          }
+          auto inputNode = n->inputs()[0]->node();
+          auto name = n->s(attr::name);
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            continue;
+          }
+
+          auto attr = attrModule.attr(name);
+          if (n->kind() == prim::GetAttr) {
+            auto type = n->output()->type();
+            // Do not record submodules. Their attributes are tracked
+            // individually.
+            if (attr.isObject() || !AliasDb::mutableType(attr.type())) {
+              continue;
+            }
+            usedAttrs.insert(attr);
+          }
+
+          if (n->kind() == prim::SetAttr || aliasDb->hasOutputWriters(n)) {
+            GRAPH_DEBUG(
+                n->kind() == prim::GetAttr ? "attribute: " + name + " in %" +
+                        n->output()->debugName() + " has inplace writer"
+                                           : "attribute: " + name + " is set");
+            insertMutableAttr(name, attr, attrModule);
+          }
+        }
+      }
+    }
+    // FIXME: Current Alias analysis fails to track subvalues.
+    // This is not a common scenario, for freezing, detect and error out.
+    for (auto it = usedAttrs.begin(); it != usedAttrs.end();) {
+      auto& val = *it;
+      it++;
+      for (auto rhs = it; rhs != usedAttrs.end(); rhs++) {
+        TORCH_CHECK(
+            !val.overlaps(*rhs),
+            "module contains attributes values that overlaps ",
+            val,
+            " and ",
+            *rhs);
+      }
+    }
+  }
+
+  // Prepraring for clean up phase. At this point, record all  subModules that
+  // contains mutable attributes.
+  void recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* subBlock : n->blocks()) {
+          blocks.push(subBlock);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto& name = n->s(attr::name);
+          if (module_.hasattr(name)) {
+            auto attr = module_.attr(name);
+            insertMutableAttr(name, attr, module_);
+          }
+        }
+      }
+    }
+  }
+
+  IValue overrideGradient(IValue attr) {
```

Comment:
Can we not just call `getSubValues()`, and iterate through tensor subvalues, in-place removing gradient ?

Reply:
Yes cleaner and it handles missing cases.

Reply:
Eliminate tensor detach?